File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/util/routeModeLabel.kt
package com.mbta.tid.mbta_app.android.util

import android.content.res.Resources
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.model.Line
import com.mbta.tid.mbta_app.model.LineOrRoute
import com.mbta.tid.mbta_app.model.Route
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.utils.RouteModeLabelType

fun routeModeLabel(resources: Resources, lineOrRoute: LineOrRoute, isOnly: Boolean = true): String =
    routeModeLabel(resources, lineOrRoute.name, lineOrRoute.type, isOnly)

fun routeModeLabel(
    resources: Resources,
    line: Line?,
    route: Route?,
    isOnly: Boolean = true,
): String = routeModeLabel(resources, line?.longName ?: route?.label, route?.type, isOnly)

fun routeModeLabel(resources: Resources, route: Route, isOnly: Boolean = true): String =
    routeModeLabel(resources, route.label, route.type, isOnly)

fun routeModeLabel(
    resources: Resources,
    name: String?,
    type: RouteType?,
    isOnly: Boolean = true,
): String {
    val label = com.mbta.tid.mbta_app.utils.routeModeLabel(name, type)
    fun localizedType(type: RouteType) = type.typeText(resources, isOnly)
    return when (label) {
        is RouteModeLabelType.NameAndType ->
            resources.getString(R.string.route_mode_label, label.name, localizedType(label.type))
        is RouteModeLabelType.NameOnly -> label.name
        is RouteModeLabelType.TypeOnly -> localizedType(label.type)
        else -> ""
    }
}

File: iosApp/iosApp/Utils/RouteModeLabel.swift
//
//  RouteModeLabel.swift
//  iosApp
//
//  Created by esimon on 8/12/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

import Shared
import SwiftUI

func routeModeLabel(lineOrRoute: LineOrRoute, isOnly: Bool = true) -> String {
    routeModeLabel(name: lineOrRoute.name, type: lineOrRoute.type, isOnly: isOnly)
}

func routeModeLabel(line: Line?, route: Route?, isOnly: Bool = true) -> String {
    routeModeLabel(name: line?.longName ?? route?.label, type: route?.type, isOnly: isOnly)
}

func routeModeLabel(route: Route, isOnly: Bool = true) -> String {
    routeModeLabel(name: route.label, type: route.type, isOnly: isOnly)
}

func routeModeLabel(name: String?, type: RouteType?, isOnly: Bool = true) -> String {
    let label = Shared.routeModeLabel(name: name, type: type)
    func typeLabel(_ type: RouteType) -> String { type.typeText(isOnly: isOnly) }

    return switch onEnum(of: label) {
    case let .nameAndType(label):
        String(format: NSLocalizedString(
            "key/route_mode_label",
            comment: """
            A route label and route type pair,
            ex 'Red Line train' or '73 bus', used in connecting stop labels
            """
        ), label.name, typeLabel(label.type))
    case let .nameOnly(label): label.name
    case let .typeOnly(label): typeLabel(label.type)
    default: ""
    }
}
